subprojects/docs/src/docs/userguide/dep-man/01-core-dependency-management/dependency_verification.adoc

```diff
@@ -0,0 +1,880 @@
+    - On artifact commons-logging-1.2.jar (commons-logging:commons-logging:1.2) in repository 'MavenRepo': checksum is missing from verification metadata.
+    - On artifact commons-logging-1.2.pom (commons-logging:commons-logging:1.2) in repository 'MavenRepo': checksum is missing from verification metadata.
+----
+
+What this indicates is that your build requires `commons-logging` when executing `compileJava`, however the verification file doesn't contain enough information for Gradle to verify the integrity of the dependencies, meaning you need to add the required information to the verification metadata file.
+
+See <<sec:troubleshooting-verification,troubleshooting dependency verification>> for more insights on what to do in this situation.
+
+=== What checksums are verified?
+
+If a dependency verification metadata files declares more than one checksum for a dependency, Gradle will verify _all of them_ and fail if _any of them fails_.
+For example, the following configuration would check both the `md5` and `sha256` checksums:
+
+[source,xml]
+----
+<component group="org.apache.pdfbox" name="pdfbox" version="2.0.17">
+   <artifact name="pdfbox-2.0.17.jar">
+      <md5 value="c713a8e252d0add65e9282b151adf6b4" origin="official site"/>
+      <sha1 value="b5c8dff799bd967c70ccae75e6972327ae640d35" origin="official site"/>
+   </artifact>
+</component>
+----
+
+There are multiple reasons why you'd like to do so:
+
+1. an official site doesn't publish _secure_ checksums (SHA-256, SHA-512) but publishes multiple insecure ones (MD5, SHA1). While it's easy to fake a MD5 checksum and hard but possible to fake a SHA1 checksum, it's harder to fake both of them for the same artifact.
+2. you might want to add generated checksums to the list above
+3. when _updating_ dependency verification file with more secure checksums, you don't want to accidentally erase checksums
+
+[[sec:signature-verification]]
+== Verifying dependency signatures
+
+In addition to <<sec:checksum-verification,checksums>>, Gradle supports verification of signatures.
+Signatures are used to assess the _provenance_ of a dependency (it tells who signed the artifacts, which usually corresponds to who produced it).
+
+While enabling signature verification usually means a higher level of security, you might want to replace checksum verification with signature verification.
+
+[WARNING]
+====
+Signatures _can_ also used to assess the integrity of a dependency similarly to checksums.
+Signatures are signatures of the _hash_ of artifacts, not artifacts themselves.
+This means that if the signature is done on an _unsafe hash_ (even SHA1), then you're not correctly assessing the _integrity_ of a file.
+For this reason, if you care about both, you need to add both signatures _and_ checksums to your verification metadata.
+====
+
+However:
+
+- Gradle only supports verification of signatures published on remote repositories as ASCII-armored PGP files
```

Comment:
What does Gradle do when the key expires?
I assume it would consider the keys to be OK for validation purposes.

Reply:
That's my assumption too. We could issue a warning that the key expired... but what would this say? Back then it wasn't expired so it's still valid...

Reply:
I'm inclined that warning is not needed (at least it should not be visible by default) because Sonatype repository is full of expired PGP keys.

I think it makes sense to document that explicitly so readers could understand what to expect rather than guess.